eng/test_tools.txt

```diff
@@ -1,15 +1,12 @@
 cryptography<4
 
 # requirements leveraged by ci for testing
-pytest==4.6.9; python_version == '2.7'
-pytest==5.4.2; python_version >= '3.5' and python_version <= '3.9'
-pytest==6.2.4; python_version >= '3.10'
-pytest-asyncio==0.12.0; python_version >= '3.5'
-pytest-cov==2.8.1
+pytest==7.3.1
+pytest-asyncio==0.21.0
```

Comment:
https://github.com/Azure/azure-sdk-for-python/pull/30159/commits/0ec907450bafbd4f10343a5d67944cf6b3b21cec bumps `pytest-asyncio` from `0.12.0` to `0.21.0`.


[`0.19.0` introduces a breaking change](https://pytest-asyncio.readthedocs.io/en/v0.21.0/reference/changelog.html#id11) that defaults its default mode to `strict`. 
 
https://pytest-asyncio.readthedocs.io/en/v0.21.0/concepts.html#test-discovery-modes

`strict` mode requires that all asyncio tests have `@pytest.mark.asyncio` and all async fixtures use `@pytest_asyncio.fixture`.

Setting `asyncio-mode` to `auto` (either in a config file or in the test invocation) is probably the way forward. It's basically is how `pytest-asyncio` used to work.

Reply:
👍 we have a conftest at `sdk/` we can add it there.

I saw your addition that changes `run` to `run-parallel`. Thanks! I was digging through their docs trying to understand why `-p` somehow wasn't applying hahaha

Reply:
There's a `setup.cfg` in the root of the repo that has a pytest section that might work too.

https://github.com/Azure/azure-sdk-for-python/blob/ebcfffdf965750fba692979a69aba1ba8e1ac458/setup.cfg#L1C1-L2

Reply:
Ah, fixing this is probably going to be more involved.

Setting `asyncio-mode` to `auto` breaks tests that use another async runtime (e.g. trio). The tests that use the TrioTransport fail in the azure-core tests. See:
* https://github.com/pytest-dev/pytest-asyncio/issues/124
* https://github.com/pytest-dev/pytest-asyncio/pull/380

It's unclear to me why this wasn't an issue before

It would appear the current options would be:


* Set `asyncio-mode` to `strict`. Would likely require a refactor of a bunch of tests to make sure `@pytest.mark.asyncio` and `@pytest_asyncio.fixture` is being applied appropriately.
* Set `asyncio-mode` to `auto`. This would require some way to override to `strict` for specific packages and/or tests that don't use asyncio, and would still need to do whatever fixes for those packages that would've been done in the previous bullet point.

Reply:
> Ah, fixing this is probably going to be more involved.
> 
> Setting `asyncio-mode` to `auto` breaks tests that use another async runtime (e.g. trio). The tests that use the TrioTransport fail in the azure-core tests. See:
> 
> * [Require explicit marks for async fixtures pytest-dev/pytest-asyncio#124](https://github.com/pytest-dev/pytest-asyncio/issues/124)
> * [opt into strict mode by default pytest-dev/pytest-asyncio#380](https://github.com/pytest-dev/pytest-asyncio/pull/380)
> 
> It's unclear to me why this wasn't an issue before
> 
> It would appear the current options would be:
> 
> * Set `asyncio-mode` to `strict`. Would likely require a refactor of a bunch of tests to make sure that the appropriate decorators (`@pytest.mark.asyncio`, `@pytest_asyncio.fixture`, trio specific ones, etc...) are being applied appropriately.
> * Set `asyncio-mode` to `auto`. This would require some way to override to `strict` for specific packages and/or tests that don't use asyncio, and would still need to do whatever fixes for those packages that would've been done in the previous bullet point.

Yeah this is unfortunate. On Monday I'll dig to see if we can upgrade pytest to get the new `pluggy`, without going all the way to the newest pytest that introduces this conflict.

Additionally, I think `tox4` re-introduces [parallelism problems](https://dev.azure.com/azure-sdk/public/_build/results?buildId=2747141&view=logs&jobId=3dc8fd7e-4368-5a92-293e-d53cefc8c4b3&j=3dc8fd7e-4368-5a92-293e-d53cefc8c4b3&t=8148717c-383a-55c2-5696-70c8097a561f) that we had entirely eliminated.

Just more stuff to deal with 😮‍💨

Reply:
In the event that juggling dependency versions doesn't work, how open are you to the idea of combing through tests to use the async decorators consistently. (At that point, I'd probably break it off into its own PR)

Reply:
I think we're ok with `strict` mode, as we have been attempting to hold the line with marking async tests with `@pytest.mark.asyncio` or `@pytest.mark.trio`.

The build _currently_ is failing for a reason that doesn't make much sense to me.

After installing `tox` locally, you can repro from `sdk/core/azure-core` via

```bash
tox run -e whl -c ../../../eng/tox/tox.ini --root . -- -sk "test_send_data"
```

This narrows it to a single test case that is the majority of the failures. For some reason, an async fixture is returning a generator, not the actual yielded value.

Reply:
Both the issues are related:

```
E       TypeError: 'coroutine' object is not callable
```

and

```
E       AttributeError: 'async_generator' object has no attribute 'send_request'

```

Reply:
Ok! I think I have an answer! We will need to make our `pytest_asyncio` and `pytest_trio` fixtures _explicit_.

Not working:

```python
@pytest.fixture
async def client(port):
    async with AsyncioRequestsTransport() as transport:
        async with AsyncTestRestClient(port, transport=transport) as client:
            yield client
```

Works:

```python
@pytest_asyncio.fixture
async def client(port):
    async with AsyncioRequestsTransport() as transport:
        async with AsyncTestRestClient(port, transport=transport) as client:
            yield client
```

Reply:
Ah I think we came to the same conclusion at the same time.

I have https://github.com/Azure/azure-sdk-for-python/pull/30159/commits/53e606867b160cb39a76d0268a5eac0345f05075 that implements this for azure-core and it seems to be working?

https://dev.azure.com/azure-sdk/public/_build/results?buildId=2765694&view=results

Reply:
🚀 🚀 🚀 

Excellent! Let's see what else blows up 👍